### ISSUE TITLE
Design to toggle "compressed/experienced developer" view of courses

### ISSUE THREAD

**Issue by morsssss** [ampproject/amp.dev#2295]
Now the courses are live on amp.dev. Yay!

These courses assume the reader may not have much experience with webdev beyond knowing some HTML and CSS. There's been some demand for a compressed version of the courses for more experienced webdevs, so these folks can take the courses more quickly and won't feel turned off. We're making the compressed versions now. They'll omit the bits that explain what an API is, what web components are, etc., Another example: they'll define the cache more quickly.

The question is: how can these be presented on amp.dev?

Here's my thought: what if each page loads up both versions, and readers can use a little button to toggle between them? Ideally we could use a cookie or query param to persist the state between pages.

Normally I like cookies, but a query param might be ideal because someone might really want to share a specific view of a specific page. I don't know which is simpler to implement for us.

Or is there an easier/prettier way to do this?

Thanks!

Looking for comments here from @pbakaus , @matthiasrohmer , @sebastianbenz, @CrystalOnScript , etc.!

**Comment by pbakaus** [comment]
I like this! We run into this issue sometimes in the rest of the docs too. We could trial this functionality in the courses, and then expand it to other docs if it works well, potentially.

@sebastianbenz @matthiasrohmer thoughts on how we'd best implement this?

**Comment by matthiasrohmer** [comment]
Good idea! The easiest way to implement this is when both versions of the document would still be maintained in the same markdown file and the differentiation between easy and advanced would only happen by some kind of tagging. Similar to what we enabled for formats like with `[filter formats="stories"]This bit is only visible for stories![/filter]`.

This exact shortcode is easily extendable with a `level` option for example: `[filter level="easy"]This won't show for pro-programmers[/filter]`.

Server-side filtering via GET-parameter would of course then work best. Saves us from all the AMP cache hurdles and make things shareable as you mentioned.

Not necessarily blocked by #1694 but this would better be implemented first as the filtering during build-time is already kinda fragile and I'd like to hold back from adding any more logic into that. But I'll make this my next task so we can also continue on this one here.

**Comment by morsssss** [comment]
That sgtm too!

I was thinking that each page would load up *both* versions, and that the user could toggle between them by clicking a button that would change state... almost like tabs. And that we'd update the URL accordingly using `pushState()`. This should be reasonable performance-wise for the courses, since both versions will probably use all the same images.

But it sounds totally reasonable to me that we'd choose the version server-side instead, and that clicking the button would load the alternate version. That would cover other cases, in which the different versions loaded a larger number of different resources.

For now, it sounds like everyone likes this idea (yay!) and we can proceed by making markdown for the advanced version. This will take us a few weeks, by which time maybe we'll able to try it on the site? 🙏

**Comment by morsssss** [comment]
We're making good progress on the compressed courses, and I've started telling people they'll be available in the near future. So I thought I'd just pop this back up in your github notifications queue!

**Comment by morsssss** [comment]
Looks like #2325 is making progress... which means this is making progress too? 😄

**Comment by matthiasrohmer** [comment]
Hehe, it does 😁Hope to get this in today, then I'll open up another PR adding at least the functionality for general/advanced courses.

**Comment by morsssss** [comment]
Amazing!

**Comment by morsssss** [comment]
Looks like #2325 is almost there...

**Comment by morsssss** [comment]
Hi, @dlemm !

Glad you'll be working on this. The compressed courses are almost ready to go... exciting!

**Comment by dlemm** [comment]
Nice, I'm looking forward to it. I am making progress, too.

**Comment by morsssss** [comment]
The markdown for the compressed version of these courses is ready. How would you like it included with the current markdown?

I imagine you'll want both of these in the same file, in each case, with some sort of demarcation?

**Comment by dlemm** [comment]
Yes, it would be great to have both of these in the same file.

We can define the default level for every file itself by declaiming so in the frontmatter 

```yaml
level: advanced|beginner
```

Parts for beginners/advanced can then be filtered by 

```md
[filter level="advanced"]
Content for advanced
[/filter]
```

Let me know if you have questions

**Comment by morsssss** [comment]
This sounds nicely consistent with the rest of the site!

I assume `beginner` content would get marked up the same way?

So a file could look like this:

```
level: advanced|beginner

[filter level="beginner"]
Hey, you're a beginner. How great is that?
[/filter]
[filter level="advanced"]
Whoa, you're so advanced!
[/filter]
```

?

And you'll provide a toggle that reloads the page with the proper query string? And... this query string will persist as they navigate through the site?

+ @mandarini

**Comment by matthiasrohmer** [comment]
I'll quickly loop into this thread 🙂Sorry Ben it took so long, actually there's still another blocker with SSR that we are working on in parallel that keeps us of from landing this feature but nevertheless I asked Dennis to actually start implementing this here.

The toggle is what Dennis will work on tomorrow, it will supply a short information what the toggle is about together with the actual UI element.

**Comment by morsssss** [comment]
Sounds great - thanks, @matthiasrohmer !
